rkt/containers.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2014 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/coreos/rocket/pkg/lock"
+)
+
+type containerHandle struct {
+	uuid       string
+	lock       *lock.DirLock
+	isExited   bool
+	isGarbage  bool
+	isDeleting bool
+}
+
+type includeMask byte
+
+const (
+	includeContainersDir includeMask = 1 << iota
+	includeGarbageDir
+)
+
+// getContainerHandles returns a slice of ContainerHandle structs representing the current state of all containers
+func getContainerHandles(include includeMask) ([]*containerHandle, error) {
+	ls, err := getContainers(include)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get containers: %v", err)
+	}
+
+	var cs []*containerHandle
+	for _, uuid := range ls {
+		state, err := getContainerHandle(uuid)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error getting state of container %s, ignoring\n", uuid)
+			continue
+		}
+		cs = append(cs, state)
+	}
+
+	return cs, nil
+}
+
+// getContainerHandle returns a ContainerHandle struct representing the given container
+func getContainerHandle(uuid string) (*containerHandle, error) {
```

Comment:
If the container with the given uuid does not exist (neither in the container dir or in the garbage dir), then the function will return nil (with the error "error opening container %q: %v", uuid, err).

So the caller will not see s.isGarbage = true. The containerHandle struct will be garbage collected.